java.base/src/main/java/java/lang/StringSupport.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * This code is based on OpenJDK source file(s) which contain the following copyright notice:
+ *
+ * ------
+ * Copyright (c) 2001, 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ * ------
+ *
+ * This file may contain additional modifications which are Copyright (c) Red Hat and other
+ * contributors.
+ */
+
+package java.lang;
+
+import java.lang.ref.ReferenceQueue;
+import java.lang.ref.WeakReference;
+import java.util.Arrays;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.qbicc.rt.annotation.Tracking;
+import org.qbicc.runtime.main.InitialHeap;
+
+/**
+ * Implementation of String.intern(); pulled to a separate class because
+ * we can't define nested classes as part of a patch.
+ */
+@Tracking("src/java.base/share/native/libjava/String.c")
+class StringSupport {
+    private static final ReferenceQueue<String> REF_QUEUE = new ReferenceQueue<>();
+    private static final ConcurrentHashMap<Key, Ref> INTERNED = new ConcurrentHashMap<>();
+
+    static String intern(String candidate) {
+        // First look to see if there is a build-time interned String
+        int btIndex = Arrays.binarySearch(InitialHeap.internedStrings, candidate);
+        if (btIndex >=0) {
+            return InitialHeap.internedStrings[btIndex];
+        }
+
+        // No build-time interned String; fallback to runtime interning
+        byte coder = candidate.coder();
+        byte[] bytes = candidate.value();
+        int hc = candidate.hashCode();
+        Key key = new Key(bytes, coder, hc);
+        for (;;) {
+            Ref strRef = INTERNED.get(key);
+            if (strRef == null) {
```

Comment:
This patch is mainly just a rename & shuffle to work around limitations in the native/patcher infrastructure, but this little snippet to deal with the initial `get` being null is a change